File: Decorator/java/DecoratorPattern.java
package Decorator.java;

/**
 * @author homo.efficio@gmail.com
 *         Created on 2017-06-18.
 */
public class DecoratorPattern {

    public static void main(String[] args) {
        AbstractPizza pizza1 = new CheeseDecorator(new ThinPizza());
        showInfo(pizza1);

        AbstractPizza pizza2 = new CheeseDecorator(new HotSauceDecorator(new CheeseDecorator(new ThinPizza())));
        showInfo(pizza2);
    }

    private static void showInfo(AbstractPizza pizza) {
        System.out.println("피자 설명: " + pizza.getDescription() + ", 가격: " + pizza.getPrice() + " 원");
    }
}

abstract class AbstractPizza {

    abstract public String getDescription();
    abstract public int getPrice();
}

class ThinPizza extends AbstractPizza {

    @Override
    public String getDescription() {
        return "씬 피자(" + getPrice() + ")";
    }

    @Override
    public int getPrice() {
        return 10;
    }
}

class CrustPizza extends AbstractPizza {

    @Override
    public String getDescription() {
        return "크러스트 피자(" + getPrice() + ")";
    }

    @Override
    public int getPrice() {
        return 15;
    }
}

abstract class PizzaDecorator extends AbstractPizza {

    protected AbstractPizza pizza;

    public PizzaDecorator(AbstractPizza pizza) {
        this.pizza = pizza;
    }

    abstract public String getDecorator();
}

class CheeseDecorator extends PizzaDecorator {

    private int price = 3;

    public CheeseDecorator(AbstractPizza pizza) {
        super(pizza);
    }

    @Override
    public String getDescription() {
        return super.pizza.getDescription() + getDecorator();
    }

    @Override
    public int getPrice() {
        return super.pizza.getPrice() + this.price;
    }

    @Override
    public String getDecorator() {
        return " | 치즈(" + this.price + ")";
    }
}

class HotSauceDecorator extends PizzaDecorator {

    private int price = 2;

    public HotSauceDecorator(AbstractPizza pizza) {
        super(pizza);
    }

    @Override
    public String getDescription() {
        return super.pizza.getDescription() + getDecorator();
    }

    @Override
    public int getPrice() {
        return super.pizza.getPrice() + this.price;
    }

    @Override
    public String getDecorator() {
        return " | 핫소스(" + this.price + ")";
    }
}

File: Decorator/python/DecoratorPattern.py
import abc


class AbstractPizza(metaclass=abc.ABCMeta):
    @abc.abstractmethod
    def get_description(self):
        raise NotImplementedError

    @abc.abstractmethod
    def get_price(self):
        raise NotImplementedError



class ThinPizza(AbstractPizza):
    def __init__(self):
        pass

    def get_description(self):
        return "씬 피자(%d)" % self.get_price()

    def get_price(self):
        return 10


class CrustPizza(AbstractPizza):
    def __init__(self):
        pass
        # AbstractPizza.__init__(self)

    def get_description(self):
        return "크러스트 피자(%d)" % self.get_price()

    def get_price(self):
        return 15



class PizzaDecorator(AbstractPizza, metaclass=abc.ABCMeta):
    def __init__(self, pizza):
        self.pizza = pizza

    @abc.abstractmethod
    def get_decorator(self):
        raise NotImplementedError


class CheeseDecorator(PizzaDecorator):
    price = 3

    def __init__(self, pizza):
        PizzaDecorator.__init__(self, pizza)

    def get_description(self):
        return self.pizza.get_description() + self.get_decorator()

    def get_price(self):
        return self.pizza.get_price() + self.price

    def get_decorator(self):
        return " | 치즈(%d)" % self.price


class HotSauceDecorator(PizzaDecorator):
    price = 2

    def __init__(self, pizza):
        PizzaDecorator.__init__(self, pizza)

    def get_description(self):
        return self.pizza.get_description() + self.get_decorator()

    def get_price(self):
        return self.pizza.get_price() + self.price

    def get_decorator(self):
        return " | 핫소스(%d)" % self.price



def run():
    def show_info(pizza):
        print("피자 설명: " + pizza.get_description() + ", 가격: %d" % pizza.get_price())

    pizza1 = CheeseDecorator(ThinPizza())
    show_info(pizza1)

    pizza2 = CheeseDecorator(HotSauceDecorator(CheeseDecorator(CrustPizza())))
    show_info(pizza2)

run()
